### DISCUSSION TITLE
Enable auxiliary fan control via software

### DISCUSSION THREAD

**Discussion by huzhekun** [OpenNeptune3D/OpenNept4une#184] [Ideas]
Since taking off the bottom panel is a required step to install this firmware, i took a look under the hood of my elegoo neptune 4 max and found that the aux fan is wired in with the main part cooling fan... via wires that splits a fan header in parallel into 2 more headers and plugs... right next to 3 empty fan headers...

The irony of this aside, I wonder if it is possible to enable Auxiliary fan control via software by taking the leads for the aux fan and plugging it into one of the empty fan headers and making the macro changes needed to enable that fan as the aux fan. I have no idea if this will work, but I will be trying this when I flash opennept4une. and due to how simple this might be if it works, i would love for this to be included in documentation as well as a potential separate code snippet for the ppl who want to make this modification.

**Comment by huzhekun** [comment]
Ok, I just flashed this firmware and plugged the aux fan into fan 3 header, and i can confirm that it works and controls the aux cooling fan separately. I have not fully installed my bottom cover yet so i will take that off again and take some pictures to illustrate what i did with the wiring. if I am missing something major please let me know

**Reply by SebNania** [reply]
Hi Huzhekun,
Could you please give us some complementary informations? have you make some picture and could you please give us some details on how to make this very cool thing append?
Do you think is there a way to do that with the elegoo stock firmware?
I search for a solution about that because this secondary fan is a good idea but in some case it cause some printing fail 
thanks a lot in advance,
Sebastien

**Reply by huzhekun** [reply]
I completely forgot about this because it seemed like the interest wasn't there, I have some pictures and I can confirm it works in the stock firmware too

**Reply by huzhekun** [reply]
Aaaaaaaand I deleted the pictures...

**Reply by SebNania** [reply]
Thank you huzhekun for your time and search for the pictures :)
I will look at this point soon, i need to understand how klipper work for fan port assignation and go forward on this subject, i will put any interesting information about that here if it could help someone.
Nice day,

**Reply by huzhekun** [reply]
@SebNania i remember plugging in the aux fan into a header labeled fan 2, and I found this online cfg script to handle m106 commands

```
# Replacement-Macro to control the Fan with M106 P2 and M107 for BambuStudio
# Macro inspired by klipper-github issue https://github.com/Klipper3d/klipper/issues/2174
[gcode_macro M106]
rename_existing: G106
gcode:
    {% if params.P is defined %}
        {% if params.P|int == 2 %}
            {% if params.S is defined %}
                {% if params.S|int == 255 %}
                    {% set realspeed = 1 %}
                {% else %}
                    {% if params.S|int == 0 %}
                        {% set realspeed = 0 %}
                    {% else %}
                        {% set realspeed = params.S|float/255 %}
                    {% endif %}
                {% endif %}
            {% else %}
                {% set realspeed = 1 %}
            {% endif %}

            SET_FAN_SPEED FAN=aux_fan SPEED={realspeed}
        {% endif %}
    {% else %}
        {% if params.S is defined %}
          G106 S{params.S}
        {% else %}
          G106 S255
        {% endif %}
    {% endif %}

[gcode_macro M107]
rename_existing: G107
gcode:
    {% if params.P is defined %}
      {% if params.P|int == 2 %}
        SET_FAN_SPEED FAN=aux_fan SPEED=0
      {% endif %}      
    {% else %}
      SET_FAN_SPEED FAN=aux_fan SPEED=0  
      G107
    {% endif %}
 ```

**Reply by huzhekun** [reply]
I also added this to my printer.cfg
```
[fan_generic aux_fan]
pin: PC9
```

**Reply by SebNania** [reply]
thank you very muche for all that it will help me.

**Comment by Spurflugzeug** [comment]
Hi, I saw your post and find it really interesting. I have a Naptune 4. Unfortunately, I don't have a free fan port, but I do have two LED ports. Do you think it would be possible to plug the fan in there and configure it?

**Reply by SebNania** [reply]
I don't know for neptune 4 but on neptune 4 max there are some free ports.
I don't think a led port could be a fan port because a led only need on/off power on the port and for fan it's a PWM output to make it turn.
But i'm not a pro in electronics.

**Reply by Spurflugzeug** [reply]
Yes, I've already considered that. 
The fan only uses two wires, so no PWM.
I think that would require four wires. 
The fan control is then, I believe, via electricity, but I'm too worried about something burning out.
Thank you for your quick reply.
